### PR TITLE
Assorted "design" documentation, and record stance on custom TLS extension APIs

### PR DESCRIPTION
**Pull request by ctz**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -351,6 +351,98 @@ don't provide additional type safety. Using the
...
+Some general concepts about how the library should fit together:
+
+- Linker friendliness
+- Small mandatory API
+- Safe and sensible defaults
+- Separation of mechanism and policy
+
+### Linker friendliness
+
+When program incorporates rustls, we should endeavour that parts that are not
```
Comment (cpu) [inline CONTRIBUTING.md:365]: editorial: "When program" -> "When _a_ program"

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -351,6 +351,98 @@ don't provide additional type safety. Using the
...
+as simple as possible.  Good models for this are `simpleclient` and
+`simpleserver`.  The purpose of this is to allow people interacting
+with the library for the first time to make progress.
+
+### Safe and sensible defaults
+
+We should be confident to make decisions on behalf of our users and
+express that in the library's defaults.  It is good to encode our experience
+of TLS in this way, and we should not require that users are experts in
+TLS to end up with a result that is working, stable, and performant.
```
Comment (cpu) [inline CONTRIBUTING.md:430]: maybe "working" -> "working securely" ? I think security should be a goal of the defaults that's mentioned explicitly.

File: rustls/src/msgs/mod.rs [hunk 3]
```diff
@@ -1,4 +1,35 @@
...
+//!                        `-._                   _.-'
+//!                            `"'--...___...--'"`
+//!
+//!  ___ ___ ___ ___  ___ ___   ___ ___  ___   ___ ___ ___ ___ ___ _  _  ___
+//! | _ ) __| __/ _ \| _ \ __| | _ \ _ \/ _ \ / __| __/ __/ __|_ _| \| |/ __|
+//! | _ \ _|| _| (_) |   / _|  |  _/   / (_) | (__| _|\__ \__ \| || .` | (_ |
+//! |___/___|_| \___/|_|_\___| |_| |_|_\\___/ \___|___|___/___/___|_|\_|\___|
+//! ```
+//!
+//! <https://langsec.org/ForWantOfANail-h2hc2014.pdf>
```
Comment (cpu) [inline rustls/src/msgs/mod.rs:32]: preach! :raised_hands:

File: rustls/src/manual/features.rs [hunk 4]
```diff
@@ -59,4 +59,39 @@ and will not support:
   path building. While dangerous, all authentication can be turned off if required --
   see the [example code](https://github.com/rustls/rustls/blob/992e2364a006b2e84a8cf6a7c3eaf0bdb773c9de/examples/src/bin/tlsclient-mio.rs#L318)_ `*`
 
+### About "custom extensions"
+
+OpenSSL allows an application to add arbitrary TLS extensions (via
+the `SSL_CTX_add_custom_ext` and associated APIs).  We don't support this,
```
Comment (cpu) [inline rustls/src/manual/features.rs:65]: editorial nit: "via the `SSL_CTX_add_custom_ext` and associated APIs" -> "via the `SSL_CTX_add_custom_ext` function and associated APIs"

File: rustls/src/manual/features.rs [hunk 5]
```diff
@@ -59,4 +59,39 @@ and will not support:
...
+with the following rationale:
+
+Such an API is limited to extensions that are quite limited in scope:
```
Comment (cpu) [inline rustls/src/manual/features.rs:68]: editorial nit: double "limited", maybe say "narrow" for the second instance?

> Such an API is limited to extensions that are quite narrow in scope:

File: rustls/src/manual/features.rs [hunk 6]
```diff
@@ -59,4 +59,39 @@ and will not support:
...
+that set of extensions.  That makes the API pretty unsafe (in the
+TLS and cryptography sense, not memory safety sense).  This could
+cause security or interop failures.
+
+Instead, we suggest that potential users of that API consider:
+
+- whether their use can fit in standard extensions such as ALPN,
+  or [ALPS][alps][^2].
+- if not, whether they can fit in a more general extension, and define
+  and standardise that in the IETF TLSWG.
```
Comment (cpu) [inline rustls/src/manual/features.rs:82]: Maybe `[IETF TLSWG](https://datatracker.ietf.org/wg/tls/charter/)`?

File: rustls/src/manual/features.rs [hunk 7]
```diff
@@ -59,4 +59,39 @@ and will not support:
...
+
+Note the above is not a guarantee or offer that rustls will implement
+any specific extensions that are standardised by the IETF TLSWG.
+It is a non-goal of this project to implement absolutely everything.
+
+For experimentation and pre-standardisation testing, we suggest
+forking rustls.
+
+See also: [golang's position on such an API][golang].
```
Comment (cpu) [inline rustls/src/manual/features.rs:91]: nit: I've been taught by the Gopher crowd that you should never say "Golang", [it's "Go"](https://go.dev/doc/faq#go_or_golang).

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -351,6 +351,98 @@ don't provide additional type safety. Using the
...
+enough to have an impact on the size of a final program.
+
+### Small mandatory API
+
+We should endeavour to keep the API for achieving the most common goals
+as simple as possible.  Good models for this are `simpleclient` and
+`simpleserver`.  The purpose of this is to allow people interacting
+with the library for the first time to make progress.
+
+### Safe and sensible defaults
```
Comment (cpu) [inline CONTRIBUTING.md:425]: Should this have some interlinking with `_05_defaults` in the manual?

**Review by cpu — COMMENTED** [review]
Thanks for getting this stuff written down. I think it looks great! Here's some initial feedback.

File: rustls/src/manual/features.rs (hunk 7 above)
Comment (ctz) [reply rustls/src/manual/features.rs:91]: the more you know!

**Review by cpu — APPROVED** [review]

**Review by djc — APPROVED** [review]
Nice!

**Pull request merged**